### DISCUSSION TITLE
prCommitsPerRunLimit is only checked after each repository

### DISCUSSION THREAD

**Discussion by Algram** [renovatebot/renovate#22174] [Suggest an Idea]
**What Renovate type are you using?**

Selfhosted renovate on GitLab
Version `19.93.2` from the `renovate:19` docker image.

**Describe the bug**
It seems that renovate is only checking if the `prCommitsPerRunLimit` is reached after a complete repository run.

**To Reproduce**
- We have three repositories to renovate and `prCommitsPerRunLimit: 2`
- The first repository only needs a single commit
- Renovate commits **1** time on the first repository
- The second repository needs 50 commits
- Renovate commits **50** times on the second repository
- The third repository needs 10 commits
- Renovate says the limit is reached and stops.

**Additional information**
Renovate should check the limit on each and every commit and not at the end of renovating a repo.

Imo this is the relevant part of the feature where it only checks after each repo:
https://github.com/renovatebot/renovate/commit/d1e7c17#diff-15f1f0b406a6fc522960ef91816a8ca2R47

**Comment by rarkins** [comment]
> It seems that renovate is only checking if the prCommitsPerRunLimit is reached after a complete repository run.

The limit should be already checked at the start of every branch: https://github.com/renovatebot/renovate/blob/409c1c5f7fa0a430e09bcaa4d7de74d6449efc3a/lib/workers/repository/process/write.ts#L38

.. and each branch should have maximum one commit.

Please try to reproduce in a public repo or repos and then save the logs at debug level. Based on your description, it should not be necessary to have multiple repos or a limit > 1. i.e. a single repo with 2 branches needing creation and a limit of 1 should be enough to reproduce.
